tests/test_attributes.py

```diff
@@ -462,7 +462,136 @@ def OLD_test_get_network_attrs(self, client, network_with_data):
         assert len(net_attrs) == 3
         assert len(net_type_attrs) == 2
 
+    def test_delete_all_duplicate_attributes(self, client, network_with_data):
 
+        duplicate_attribute = JSONObject({'name': 'duplicate', 'dimension_id': None})
+
+        #use dedicated testing function  which allows duplicates
+        dupe_attr_1 = client.add_attribute_no_checks(duplicate_attribute)
+        dupe_attr_2 = client.add_attribute_no_checks(duplicate_attribute)
+
+
+        all_attrs = client.get_attributes()
+
+        assert dupe_attr_1.id in [a.id for a in all_attrs]
+        assert dupe_attr_2.id in [a.id for a in all_attrs]
+
+        #add duplicate resource attributes
+        client.add_resource_attribute('NETWORK', network_with_data.id, dupe_attr_1.id, 'Y')
+        client.add_resource_attribute('NETWORK', network_with_data.id, dupe_attr_2.id, 'Y')
+
+        #check the dupes are there
+        updated_net = client.get_network(network_with_data.id)
+        updated_net_ras = [ra.attr_id for ra in updated_net.attributes]
+        assert dupe_attr_1.id in updated_net_ras
+        assert dupe_attr_2.id in updated_net_ras
+
+        #now add duplicate attrs to the template type
+        templatetype_to_update = network_with_data.types[0].id
+        client.add_typeattr(JSONObject({'type_id': templatetype_to_update,
+                                        'attr_id': dupe_attr_1.id}))
+        client.add_typeattr(JSONObject({'type_id': templatetype_to_update,
+                                        'attr_id': dupe_attr_2.id}))
+
+        #check the dupes are there
+        updated_type = client.get_templatetype(templatetype_to_update)
+        assert dupe_attr_1.id in [ta.attr_id for ta in updated_type.typeattrs]
+        assert dupe_attr_1.id in [ta.attr_id for ta in updated_type.typeattrs]
+
+        client.delete_all_duplicate_attributes()
+
+        #check the dupes are gone
+        updated_net = client.get_network(network_with_data.id)
+        updated_net_ras = [ra.attr_id for ra in updated_net.attributes]
+        assert dupe_attr_1.id in updated_net_ras
+        assert dupe_attr_2.id not in updated_net_ras
+
+        #check the dupes are gone
+        updated_type = client.get_templatetype(templatetype_to_update)
+        assert dupe_attr_1.id in [ta.attr_id for ta in updated_type.typeattrs]
+        assert dupe_attr_2.id not in [ta.attr_id for ta in updated_type.typeattrs]
+
+        reduced_attrs = client.get_attributes()
+
+        #check that the first attr is there, but the dupe is not.
+        assert dupe_attr_1.id in [a.id for a in reduced_attrs]
+        assert dupe_attr_2.id not in [a.id for a in reduced_attrs]
```

Comment:
Is it certain that `delete_all_duplicate_attributes()` will prefer to delete the second attr rather than the first?  Is this due to the lower `id`?

Reply:
Yes, I've chosen to always keep the attribute with the lowest ID. This is a somewhat arbitrary metric for deciding which attribute to keep but I chose it primarily because this is likely to be linked to more foreign keys than the others. 

Does that explain it?

Reply:
Yes, and I agree with the rationale for choosing the lowest id discussed above in `delete_duplicate_attributes` (and see the note about a return value there).

For extra clarity, it might be useful to explicitly identify the attr with the lower of the two ids here with a `min` and then the assert on 517-8 becomes a test that the lower is preserved while the other isn't. At present it's a test that the first passed to `add_attribute_no_checks()` is preserved, which happens also to be the lower id.